src/libponyc/codegen/genopt.cc

```diff
@@ -528,380 +513,370 @@ class DispatchPonyCtx : public FunctionPass
-  }
-
-  Value* mergeConstant(IRBuilder<>& builder, Value* ctx, size_t size)
-  {
-    Function* used_alloc_fn;
-    IntegerType* int_type;
-
-    if(size == 0)
-    {
-      return ConstantPointerNull::get(builder.getInt8PtrTy());
-    } else if(size <= HEAP_MAX) {
-      used_alloc_fn = alloc_small_fn;
-      int_type = unwrap<IntegerType>(c->i32);
-      size = ponyint_heap_index(size);
-    } else {
-      used_alloc_fn = alloc_large_fn;
-      int_type = unwrap<IntegerType>(c->intptr);
-    }
-
-    Value* args[2];
-    args[0] = ctx;
-    args[1] = ConstantInt::get(int_type, size);
-
-    CallInst* inst = builder.CreateCall(used_alloc_fn,
-      ArrayRef<Value*>(args, 2));
-    inst->setTailCall();
-    return inst;
-  }
-
-  Function* declareAllocFunction(Module& m, std::string const& name,
-    Type* size_type, size_t min_size, bool can_be_null)
-  {
-    Type* params[2];
-    params[0] = unwrap(c->void_ptr);
-    params[1] = size_type;
-
-    FunctionType* fn_type = FunctionType::get(unwrap(c->void_ptr),
-      ArrayRef<Type*>(params, 2), false);
-    Function* fn = Function::Create(fn_type, Function::ExternalLinkage, name,
-      &m);
-
-    fn->addFnAttr(Attribute::NoUnwind);
-    fn->addFnAttr(Attribute::InaccessibleMemOrArgMemOnly);
-
-    AttrBuilder attrs(m.getContext());
-    if(can_be_null)
-      attrs.addDereferenceableOrNullAttr(min_size);
-    else
-      attrs.addDereferenceableAttr(min_size);
-    attrs.addAlignmentAttr(target_is_ilp32(c->opt->triple) ? 4 : 8);
-    fn->addRetAttrs(attrs);
-
-    return fn;
-  }
-
-  bool isZeroRealloc(CallInst* realloc)
-  {
-    Value* new_size = realloc->getArgOperand(2);
-    ConstantInt* new_int_size = dyn_cast_or_null<ConstantInt>(new_size);
-    if(new_int_size == NULL)
-      return false;
-
-    uint64_t new_alloc_size = new_int_size->getZExtValue();
-    if(new_alloc_size != 0)
-      return false;
-    return true;
-  }
-};
-
-char MergeRealloc::ID = 2;
-
-static RegisterPass<MergeRealloc>
-  MR("mergerealloc", "Merge successive reallocations of the same variable");
-
-static void addMergeReallocPass(const PassManagerBuilder& pmb,
-  PassManagerBase& pm)
-{
-  if(pmb.OptLevel >= 2)
-    pm.add(new MergeRealloc());
-}
-
-class MergeMessageSend : public FunctionPass
+// This optimization pass is not safe...
+//
+// It tries to remove calls to `pony_realloc` by making the original pointer
+// bigger to match the later realloc size, but it doesn't do any analysis
+// to prove that the reallocation didn't have some semantic importance.
+// For example, when using `Array.chop` or `String.chop` followed by a `push`
+// onto the left side of the chop point, the left side must be reallocated
+// in order to avoid mutating the right side. But this pass doesn't do any
+// analysis to check if the pointer may possibly be accessible from some SSA,
+// so it doesn't know that removing the reallocation will break the semantics.
+//
+// class MergeRealloc : public PassInfoMixin<MergeRealloc>
```

Comment:
@kulibali @SeanTAllen - This optimization pass isn't safe, and the work to make it safe would be very non-trivial (if it's possible at all).

How do you want me to handle this? Should I leave it in the codebase in a disabled state with a note about it needing to be re-designed?

Or should I just go ahead and remove it entirely, and expect that anyone who wants to work on it can dig it up in history if they want to use the old code as a starting point?

If it were entirely up to me, I'd do the latter. But I wanted to check in for your thoughts.

Reply:
I went ahead and did the latter, leaving a release note as a trail.